sky/serve/load_balancer.py

```diff
@@ -18,25 +21,135 @@
+# be restarted to take up configuration changes.  When done naively (as the
+# PR currently does), restarting will cause all existing connections to be
+# dropped, forcing clients to retry later.  As written, this means that any
+# time a replica is added or removed, connections to all replicas will be
+# impacted.  There are two ways to deal with this:
+#   - Hot Restart:  Envoy has a mechanism for "hot restart".   When used, a new
+#   Envoy boots and begins handling connections with the new coniguration while
+#   the old one drains.
+#   - XDS:  Envoy has a robust remote update protocol that allows changes to be
+#   made to it's configuration at runtime.  XDS support may be needed anyways
+#   if we move to having multiple Envoy Load balancers that are not necessarily
+#   co-located with the controller.  As a result it may be worth skipping hot
+#   restart and building this out instead.
+#
+# - Autoscaling:  PythonLoadBalancer uses a list of each request and the
+# time it was handled to calculate a qps number which the controller uses to
+# make autoscaling decisions.  Envoy doesn't export stats with this level of
+# detail, so we will need to find a different approach.  A couple of
+# considerations:
+#    - We could simulate the existing functionality by simply asking Envoy how
+#    many requests it has handled for each replica on some regular cadence, and
+#    using that to calculate an estimated qps number.   This should be
+#    relatively straight forward to implement.
+#    - We should consider, however, whether qps is really the right way to be
+#    making auto-scaling decisions.  For AI serving workloads, does qps really
+#    tell you how loaded the replica is?   In the non-AI world, one would
+#    generally use CPU/Memory pressure for this purpose, however there are some
+#    challenges here with regard to measuring load on GPUs.  This requires
+#    further discussion.
+#
+# - Single Envoy:  In this PR we boot an Envoy per service, mostly to be
+# consistent with the approach the PythonLoadBalancer takes.   This approach
+# has two major problems (1) given that we could have several Envoy Instances,
+# we need to be fairly stingy with how much CPU/Memory we give each one.  (2)
+# This approach will significantly complicate things for Kubernets Support
+# (see below).
+#
+# - Kubernetes Support:  The PR relies on using the `docker` command line to
+# boot an Envoy container per service.  This won't work in Kubernetes where the
+# sky controller is a pod, and thus can't run sub-containers.   There are two
+# ways this could be handled.  1) Run envoy as a process instead of a docker
+# container.  2) Switch to single-envoy per controller, and boot that Envoy in
+# the pod-spec of the Kubernetes controller.
+#
+# - Weighted Least Request: We should used the weighted least request algorithm
+# instead of round robin.  The switch shouldn't be hard, just haven't done it
+# yet.
```

Comment:
So Envoy has basically three ways to configure this:

1) Envoy has a bunch of load balancing algorithms built in: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers

Note by tweaking things like weights, I suspect a large fraction of interesting things people may want to be do can be simulated.

2) Envoy has WASM support.  This allows users to plug in custom code to do all sorts of things at runtime.   In the very long term, I could imagine people writing python programs that compile to wasm for this case.

3) Envoy has a robust mechanism for installing custom c++ modules that can do more-or-less whatever you want.

Realistically (2) and (3) are probably too heavy a lift for a typical sky pilot user, but it's something the project can consider if we really need it.

I have no issue with maintaining the python load balancer.   If we want to maintain both long term, it may impact the architecture we end up with to some degree.  For example, instead of having controllers speak xDS to remote Envoys directly, we may want to come up with a simplified protocol for use within the project, that's converted to xDS locally in the Envoy handling code.

Reply:
Also could we talk out of band at some point about *how* users want to customize the load balancing logic? I'd like to start thinking about it.  Like what are they trying to achieve? What's the use case?